vtex-sites/nextjs.store
# feat: Add instant CMS Preview

## What's the purpose of this pull request?
This PR adds instant preview capabilities to this starter by following Next.JS's own guide on using the [preview mode](https://nextjs.org/docs/advanced-features/preview-mode)

The content edition flow can be seen on the video below
![cms2](https://user-images.githubusercontent.com/1753396/192648673-c75aca27-621f-448c-8f69-4d6e30aac7c9.gif)

> Note that CMS API is currently SUPER slow, so UX may be REALLY BAD

## How does it work?
This PR adds a new function called `/api/preview`. This function implements the same algorithm described on NextJS docs, however, two things are still missing:
1. Be able to preview only on .vtex.app or when a logged user is viewing the content
2. Be able to redirect the user to the right content for previewing.

`1` is desirable but I don't think it's required because, even though an attacker can fetch /api/preview in production, the cracker would still need the content's UUID and version, a highly unlikeable situation. If an attacker has these info, we should be more worried, because this means they've broken into the admin somehow.

`2` is also desirable but I think can be done in another PR when we actually have multiple content-types. Currently, we only have the home page.

## How to test it?
Open the [CMS config](https://storeframework.myvtex.com/admin/apps/vtex.admin-cms-graphql-rc@1.5.0/setup/) and set this deploy preview as the preview url:
<img width="1037" alt="image" src="https://user-images.githubusercontent.com/1753396/192643509-47340fe4-3f85-44d7-8825-d4d0a7256532.png">

Save and than open the [CMS page](https://storeframework.myvtex.com/admin/new-cms/faststore). Select a content type you would like to change (home page in our case), edit some things and save. By this point, you should be able to see the preview button:
<img width="1677" alt="image" src="https://user-images.githubusercontent.com/1753396/192644272-74371870-c315-4aa1-875d-fc0b3daf9e42.png">

Click on the preview. You should be taken to the same url you registered on the CMS config with the content being previewed without any cache
<img width="1680" alt="image" src="https://user-images.githubusercontent.com/1753396/192644578-3dbe28ee-c713-4d9a-8cab-0b3e0e28cfb9.png">

> Note that for previewing, a cookie will be set on your browser. After finish editing the CMS, you can clear the cookies to start seeing the production website.

## References
- [NextJS preview mode](https://nextjs.org/docs/advanced-features/preview-mode)
